Add create_distributed_restore_point() description (#673)
timescale/docs.timescale.com-content: api.md

CALL distributed_exec('CREATE DATABASE dist_database', transactional => FALSE);
```

---
## create_distributed_restore_point() [](create_distributed_restore_point)

Create a named point for performing restore across a distributed database.

> This function is an analogue of `pg_create_restore_point()` designed to work with a distributed database

Reviewer: Can you add a link to PG documentation, which describes the function?

and taking into account synchronization of the remote transactions.

> 

Reviewer: I think it is important to describe what it does:
- It will obtain locks and lock all distributed hypertables for queries to have a consistent state at the restore point.
- It will create restore point on every data node and the access node with the given name.

Author: I guess I can add more info about it, but just to note it does not lock any hypertables at all. It is more about remote transaction synchronization and postponing `COMMIT PREPARE` command in some cases to ensure consistency. Not really sure that this implementation details should be exposed here.

Reviewer: Sorry, I believe I wasn't correct in the formulation. I feel it will be good for users to know that the distributed restore point is blocking, so uncommitted distributed transactions will wait until the function completes. Do I follow correctly?

Author: It does not block remote transactions completely, only when they decide it is time to commit. This commit operation might be postponed until the function completes, in order to achieve consistent WAL history across the cluster. I don't really think this should be mentioned in more details then necessary in the API description, unless people disagree.

Reviewer: I see the improvement. I think it is good.

#### Required Arguments [](create_distributed_restore_point-required-arguments)

|Name|Description|
|---|---|
| `name` | The restore point name. |

#### Returns

| Column          | Description                               |
|-----------------|-------------------------------------------|
| `node_name`     | Node name (or `NULL` for access node).    |
| `node_type`     | Node type (`access_node` or `data_node`). |

> | `restore_point` | Restore point LSN.                        |

Reviewer: Is it possible to add link to a PG page describing LSN?

#### Sample Usage [](create_distributed_restore_point-examples)

Create the restore point `pitr` across three data nodes and access node:

```sql
SELECT * FROM create_distributed_restore_point('pitr');
 node_name |  node_type  | restore_point 
-----------+-------------+---------------
           | access_node | 0/3694A30
 dn1       | data_node   | 0/3694A98
 dn2       | data_node   | 0/3694B00
 dn3       | data_node   | 0/3694B68
(4 rows)
```

> 

Reviewer: It is good to describe all cases when an error will happen.

Author: I've updated the description with requirements, not really sure what other errors are expected to be described here

Reviewer: According to the implementation there are many cases when error will be returned. See https://github.com/timescale/timescaledb/pull/2843/files#diff-cf55e2617d97e45d62909cc52c6382e2ae31e0cb57e40e246638c73036f6d057R77-R114

---

## create_distributed_restore_point() [](create_distributed_restore_point)

Create a named point for performing restore across a distributed database.

> This function is an analogue of [pg_create_restore_point](https://www.postgresql.org/docs/current/functions-admin.html) designed

Reviewer: I still think it is good to put the function name into code as in the rest of the doc:
```suggestion
This function is an analogue of [`pg_create_restore_point`](https://www.postgresql.org/docs/current/functions-admin.html) designed
```

Author: You sure we need that change? it doesn't make much sense from markdown perspective, since link will be highlighted anyway. Also couldn't find any similar way of doing as you suggested in the API doc, do you have any reference?

Reviewer: It does. See an example in the first sentence under https://docs.timescale.com/latest/api#attach_data_node

> This function is an analogue of [pg_create_restore_point()](https://www.postgresql.org/docs/current/functions-admin.html) designed

Reviewer: The rest of the API doc doesn't use function parentheses after the function name during referring.
You can also use code inside reference. I suggest:
```suggestion
This function is an analogue of [`pg_create_restore_point`](https://www.postgresql.org/docs/current/functions-admin.html) designed
```

to work with a distributed database.

On successful execution the same named restore point will be created on the access node and each
associated data node. It synchronizes execution of ongoing remote transactions to ensure that
each node will have a consistent restore point view across the cluster.

This function can only be run on the access node and requires superuser preveleges.

#### Required Arguments [](create_distributed_restore_point-required-arguments)

|Name|Description|
|---|---|
| `name` | The restore point name. |

#### Returns [](create_distributed_restore_point-returns)

| Column          | Description                               |
|-----------------|-------------------------------------------|
| `node_name`     | Node name (or `NULL` for access node)    |
| `node_type`     | Node type (`access_node` or `data_node`) |

> | `restore_point` | Restore point [LSN](https://www.postgresql.org/docs/9.4/datatype-pg-lsn.html) |

Reviewer: You don't want to refer version 9.4 of docs.
```suggestion
| `restore_point` | Restore point [LSN](https://www.postgresql.org/docs/current/datatype-pg-lsn.html) |
```

---

## create_distributed_restore_point() [](create_distributed_restore_point)

> Create a named point for performing restore across a distributed database.

Reviewer 2: Consistency, do we say "distributed database" or "multi-node TimescaleDB cluster"?

This function is an analogue of [`pg_create_restore_point`](https://www.postgresql.org/docs/current/functions-admin.html) designed
to work with a distributed database.

> On successful execution the same named restore point will be created on the access node and each

Reviewer 2: ```suggestion
On successful execution, the same named restore point will be created on the access node and on each
```

associated data node. It synchronizes execution of ongoing remote transactions to ensure that

> each node will have a consistent restore point view across the cluster.

Reviewer 2: Add something why you want this?  e.g.,:

"With such a restore point, and backup/restore point should return to the same restore point across nodes in a multi-node TimescaleDB cluster, to ensure a consistent point in time for recovery."

> This function can only be run on the access node and requires superuser preveleges.

Reviewer 2: ```suggestion
This function can only be run on the access node and requires superuser privileges.
```

#### Required Arguments [](create_distributed_restore_point-required-arguments)

|Name|Description|
|---|---|
| `name` | The restore point name. |

#### Errors

An error will be given if:
- Restore point `name` is more than 64 chars.
- Recovery is in progress.
- Current WAL level is not set to `replica` or `logical`.
- Current user is not a superuser.
- Current server is not the access node.

> - Timescaledb 2PC transactions are disabled.

Reviewer: The correct name is `TimescaleDB`:
```suggestion
- TimescaleDB's 2PC transactions are disabled.
```

#### Returns [](create_distributed_restore_point-returns)

| Column          | Description                               |
|-----------------|-------------------------------------------|
| `node_name`     | Node name (or `NULL` for access node)    |
| `node_type`     | Node type (`access_node` or `data_node`) |

> | `restore_point` | Restore point [LSN](https://www.postgresql.org/docs/current/datatype-pg-lsn.html) |

Reviewer 2: Describe the type of returned values?

---

## create_distributed_restore_point() [](create_distributed_restore_point)

> Create a named point for performing restore across a multi-node TimescaleDB cluster.

Reviewer 3: Make it similar to the PG documentation, but adapt to distributed case:

```suggestion
Create a named marker record in the write-ahead logs of all nodes in a multi-node TimescaleDB cluster. This named marker, i.e., "restore point", can later be used as a recovery target, ensuring the entire multi-node cluster can be restored to a consistent state. The function returns the write-ahead log locations for all nodes where the marker record was written.
```

> This function is an analogue of [`pg_create_restore_point`](https://www.postgresql.org/docs/current/functions-admin.html) designed

Reviewer 3: I suggest using the link to the table where the function is specified instead of the top-level page: 

```suggestion
This function is an analogue of [`pg_create_restore_point`](https://www.postgresql.org/docs/current/functions-admin.html#FUNCTIONS-ADMIN-BACKUP-TABLE) designed
```

to work with a distributed database.

On successful execution, the same named restore point will be created on the access node and on each
associated data node. It synchronizes execution of ongoing remote transactions to ensure that
each node will have a consistent restore point view across the cluster.

With such a restore point, and backup/restore point should return to the same restore point

> across nodes in a multi-node TimescaleDB cluster, to ensure a consistent point in time for recovery.

Author: I've replaced it with the paragraph that Erik suggested below
